### ISSUE TITLE
Cilium on EKS continually creates network interfaces

### ISSUE THREAD

**Issue by ScottKGregory** [cilium/cilium#15877]
## Bug report

After upgrading Cilium version to 1.9.6 via helm using the config below, the operator has started to log the same few warning messages constantly and is continually creating network interfaces within AWS (~2500 in 4 days). The cluster is not running at capacity and has no pods stuck etc. all are running.

*Helm config*
```yaml
egressMasqueradeInterfaces: eth0
eni: true
ipam:
  mode: eni
nodeinit:
  enabled: true
tunnel: disabled
operator:
  enabled: true
  nodeSelector:
    kubernetes.io/os: linux
    kubernetes.io/arch: amd64
```

*Operator Logs*
```
level=info msg=""Synchronized ENI information"" numInstances=5 numSecurityGroups=10 numSubnets=6 numVPCs=1 subsys=eni"
level=info msg=""Resolving IP deficit of node"" available=15 availableForAllocation=1 availableInterfaces=3 instanceID=i-097df4c471e14fae8 maxIPsToAllocate=3 name=ip-172-18-147-43.eu-west-2.compute.internal neededIPs=3 remainingInterfaces=3 selectedInterface=eni-0aa4c88d9c9e96c28 selectedPoolID=subnet-0356b39262ec75868 subsys=ipam used=15"
level=info msg=""Resolving IP deficit of node"" available=15 availableForAllocation=1 availableInterfaces=3 instanceID=i-01315b97b6f3a25e2 maxIPsToAllocate=3 name=ip-172-18-156-35.eu-west-2.compute.internal neededIPs=3 remainingInterfaces=3 selectedInterface=eni-04fa74ffcda94c138 selectedPoolID=subnet-0e12ca8e3c344760f subsys=ipam used=13"
level=info msg=""No more IPs available, creating new ENI"" addresses=3 instanceID=i-097df4c471e14fae8 name=ip-172-18-147-43.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-0e12ca8e3c344760f subsys=ipam"
level=warning msg=""Unable to assign additional IPs to interface, will create new interface"" error=""PrivateIpAddressLimitExceeded: Number of private addresses will exceed limit.\n\tstatus code: 400, request id: f81a78dc-2c31-46ba-b6eb-ccb1b27288b4"" instanceID=i-097df4c471e14fae8 ipsToAllocate=1 name=ip-172-18-147-43.eu-west-2.compute.int
level=warning msg=""Unable to assign additional IPs to interface, will create new interface"" error=""PrivateIpAddressLimitExceeded: Number of private addresses will exceed limit.\n\tstatus code: 400, request id: d503904d-56bb-404e-a887-258f95bf873a"" instanceID=i-01315b97b6f3a25e2 ipsToAllocate=1 name=ip-172-18-156-35.eu-west-2.compute.int
level=warning msg=""Unable to assign additional IPs to interface, will create new interface"" error=""PrivateIpAddressLimitExceeded: Number of private addresses will exceed limit.\n\tstatus code: 400, request id: 5a7ec9cf-4761-4751-8298-d87358a35fbe"" instanceID=i-081fc9fa7a06238e7 ipsToAllocate=1 name=ip-172-18-124-212.eu-west-2.compute.in
level=info msg=""Created new ENI"" addresses=3 eniID=eni-0be5758d0970f8435 instanceID=i-097df4c471e14fae8 name=ip-172-18-147-43.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-0e12ca8e3c344760f subsys=ipam"
level=info msg=""Created new ENI"" addresses=3 eniID=eni-06b9a38175f192ef9 instanceID=i-081fc9fa7a06238e7 name=ip-172-18-124-212.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-05790bb4363d680a7 subsys=ipam"
level=warning msg=""Unable to assign additional IPs to interface, will create new interface"" error=""PrivateIpAddressLimitExceeded: Number of private addresses will exceed limit.\n\tstatus code: 400, request id: 34c8b59d-cc19-4ec3-910f-fbd4cb4a5977"" instanceID=i-081fc9fa7a06238e7 ipsToAllocate=1 name=ip-172-18-124-212.eu-west-2.compute.in
level=info msg=""No more IPs available, creating new ENI"" addresses=3 instanceID=i-01315b97b6f3a25e2 name=ip-172-18-156-35.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-0e12ca8e3c344760f subsys=ipam"
level=info msg=""Created new ENI"" addresses=3 eniID=eni-025701a325f3c92b7 instanceID=i-01315b97b6f3a25e2 name=ip-172-18-156-35.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-0e12ca8e3c344760f subsys=ipam"
level=info msg=""Created new ENI"" addresses=3 eniID=eni-0175415895e16af20 instanceID=i-097df4c471e14fae8 name=ip-172-18-147-43.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-0e12ca8e3c344760f subsys=ipam"
level=warning msg=""Unable to create interface on instance: unable to attach ENI at index 3: AttachmentLimitExceeded: Interface count 4 exceeds the limit for t3.medium\n\tstatus code: 400, request id: 5689426a-4b6a-4b08-8bbf-9309700d1558"" instanceID=i-097df4c471e14fae8 name=ip-172-18-147-43.eu-west-2.compute.internal subsys=ipam"
level=warning msg=""Unable to maintain ip pool of node"" error=""unable to attach ENI at index 3: AttachmentLimitExceeded: Interface count 4 exceeds the limit for t3.medium\n\tstatus code: 400, request id: 367a8952-db2e-4b2c-bc27-5beefbac67df"" instanceID=i-01315b97b6f3a25e2 name=ip-172-18-156-35.eu-west-2.compute.internal subsys=ipam"
level=warning msg=""Unable to maintain ip pool of node"" error=""unable to attach ENI at index 3: AttachmentLimitExceeded: Interface count 4 exceeds the limit for t3.medium\n\tstatus code: 400, request id: b1487864-7165-418f-8890-6194a7a81460"" instanceID=i-0b8c0cd72dbfd2685 name=ip-172-18-105-176.eu-west-2.compute.internal subsys=ipam"
level=warning msg=""Unable to create interface on instance: unable to attach ENI at index 3: AttachmentLimitExceeded: Interface count 4 exceeds the limit for t3.medium\n\tstatus code: 400, request id: 367a8952-db2e-4b2c-bc27-5beefbac67df"" instanceID=i-01315b97b6f3a25e2 name=ip-172-18-156-35.eu-west-2.compute.internal subsys=ipam"
level=warning msg=""Unable to maintain ip pool of node"" error=""unable to attach ENI at index 3: AttachmentLimitExceeded: Interface count 4 exceeds the limit for t3.medium\n\tstatus code: 400, request id: 5689426a-4b6a-4b08-8bbf-9309700d1558"" instanceID=i-097df4c471e14fae8 name=ip-172-18-147-43.eu-west-2.compute.internal subsys=ipam"
level=info msg=""Resolving IP deficit of node"" available=15 availableForAllocation=1 availableInterfaces=3 instanceID=i-01315b97b6f3a25e2 maxIPsToAllocate=3 name=ip-172-18-156-35.eu-west-2.compute.internal neededIPs=3 remainingInterfaces=3 selectedInterface=eni-04fa74ffcda94c138 selectedPoolID=subnet-0e12ca8e3c344760f subsys=ipam used=13"
level=info msg=""No more IPs available, creating new ENI"" addresses=3 instanceID=i-0b8c0cd72dbfd2685 name=ip-172-18-105-176.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-05790bb4363d680a7 subsys=ipam"
level=info msg=""Resolving IP deficit of node"" available=15 availableForAllocation=1 availableInterfaces=3 instanceID=i-081fc9fa7a06238e7 maxIPsToAllocate=3 name=ip-172-18-124-212.eu-west-2.compute.internal neededIPs=3 remainingInterfaces=3 selectedInterface=eni-0aa08b1c11119ca4c selectedPoolID=subnet-0f5ec4805aaaf2c0a subsys=ipam used=12"
level=warning msg=""Unable to assign additional IPs to interface, will create new interface"" error=""PrivateIpAddressLimitExceeded: Number of private addresses will exceed limit.\n\tstatus code: 400, request id: 61517681-44df-4ba2-a00d-fbfca6a0dd90"" instanceID=i-097df4c471e14fae8 ipsToAllocate=1 name=ip-172-18-147-43.eu-west-2.compute.int
level=info msg=""No more IPs available, creating new ENI"" addresses=3 instanceID=i-097df4c471e14fae8 name=ip-172-18-147-43.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-0e12ca8e3c344760f subsys=ipam"
level=info msg=""No more IPs available, creating new ENI"" addresses=3 instanceID=i-081fc9fa7a06238e7 name=ip-172-18-124-212.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-05790bb4363d680a7 subsys=ipam"
level=info msg=""No more IPs available, creating new ENI"" addresses=3 instanceID=i-0b8c0cd72dbfd2685 name=ip-172-18-105-176.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-05790bb4363d680a7 subsys=ipam"
level=info msg=""Created new ENI"" addresses=3 eniID=eni-05922ab312a40c5ab instanceID=i-081fc9fa7a06238e7 name=ip-172-18-124-212.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-05790bb4363d680a7 subsys=ipam"
level=info msg=""Created new ENI"" addresses=3 eniID=eni-050696d7be78e31ea instanceID=i-097df4c471e14fae8 name=ip-172-18-147-43.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-0e12ca8e3c344760f subsys=ipam"
level=info msg=""No more IPs available, creating new ENI"" addresses=3 instanceID=i-01315b97b6f3a25e2 name=ip-172-18-156-35.eu-west-2.compute.internal securityGroupIDs=""[sg-0ba2842737112d374]"" subnetID=subnet-0e12ca8e3c344760f subsys=ipam"
level=warning msg=""Unable to create interface on instance: unable to attach ENI at index 3: AttachmentLimitExceeded: Interface count 4 exceeds the limit for t3.medium\n\tstatus code: 400, request id: 90f15fa7-c3d8-4eeb-bbb8-45e1ef82b74d"" instanceID=i-097df4c471e14fae8 name=ip-172-18-147-43.eu-west-2.compute.internal subsys=ipam"
level=warning msg=""Unable to maintain ip pool of node"" error=""unable to attach ENI at index 3: AttachmentLimitExceeded: Interface count 4 exceeds the limit for t3.medium\n\tstatus code: 400, request id: 90f15fa7-c3d8-4eeb-bbb8-45e1ef82b74d"" instanceID=i-097df4c471e14fae8 name=ip-172-18-147-43.eu-west-2.compute.internal subsys=ipam"
level=warning msg=""Unable to maintain ip pool of node"" error=""unable to attach ENI at index 3: AttachmentLimitExceeded: Interface count 4 exceeds the limit for t3.medium\n\tstatus code: 400, request id: 7ce79ea4-ad6f-473e-b85f-326ef043bef8"" instanceID=i-0b8c0cd72dbfd2685 name=ip-172-18-105-176.eu-west-2.compute.internal subsys=ipam"
level=warning msg=""Unable to maintain ip pool of node"" error=""unable to attach ENI at index 3: AttachmentLimitExceeded: Interface count 4 exceeds the limit for t3.medium\n\tstatus code: 400, request id: ecfa46b9-0516-4f8e-bdad-bddf011f8b82"" instanceID=i-081fc9fa7a06238e7 name=ip-172-18-124-212.eu-west-2.compute.internal subsys=ipam"
level=warning msg=""Unable to maintain ip pool of node"" error=""unable to attach ENI at index 3: AttachmentLimitExceeded: Interface count 4 exceeds the limit for t3.medium\n\tstatus code: 400, request id: c23a005a-0222-47ea-b084-b0c0527d8752"" instanceID=i-01315b97b6f3a25e2 name=ip-172-18-156-35.eu-west-2.compute.internal subsys=ipam"
```

**General Information**

- Cilium version (run `cilium version`)
  - Client: 1.9.6 38dd27a 2021-04-20T14:57:11-07:00 go version go1.15.11 linux/amd64
  - Daemon: 1.9.6 38dd27a 2021-04-20T14:57:11-07:00 go version go1.15.11 linux/amd64
- Kernel version (run `uname -a`)
  - `Linux ip-172-18-105-176.eu-west-2.compute.internal 5.4.105-48.177.amzn2.x86_64 #1 SMP Tue Mar 16 04:56:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
- Orchestration system version in use (e.g. `kubectl version`, ...)
  - Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"clean", BuildDate:"2021-02-18T16:12:00Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"windows/amd64"}
  - Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.6-eks-49a6c0", GitCommit:"49a6c0bf091506e7bafcdb1b142351b69363355a", GitTreeState:"clean", BuildDate:"2020-12-23T22:10:21Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
  - Helm Version: version.BuildInfo{Version:"v3.5.3", GitCommit:"041ce5a2c17a58be0fcd5f5e16fb3e7e95fea622", GitTreeState:"dirty", GoVersion:"go1.15.8"}
- Link to relevant artifacts (policies, deployments scripts, ...)
- Generate and upload a system zip:
```
curl -sLO https://git.io/cilium-sysdump-latest.zip && python cilium-sysdump-latest.zip
```
[cilium-sysdump-20210427-112044.zip](https://github.com/cilium/cilium/files/6383777/cilium-sysdump-20210427-112044.zip)

**How to reproduce the issue**

1. Use helm to upgrade Cilium to release 1.9.6 in an EKS cluster using the helm config above
2. Observe the log messages and network interface creation

**Comment by ScottKGregory** [comment]
As a stop-gap measure we've rolled back to v1.9.5 which has indeed stopped the warnings and network interface creation issues.

**Comment by dctrwatson** [comment]
I also saw this same behavior after upgrading to 1.9.6. It was more surprising because we limit ENIs too.

This works around the issue in 1.9.6 (while limiting to only 2 secondary ENIs):
```
aws-instance-limit-mapping: |
{
"c5.large":     "3, 9, 9",
"c5.xlarge":    "3, 14, 14",
"c5.2xlarge":   "3, 14, 14",
"c5.4xlarge":   "3, 29, 29",
"c5.9xlarge":   "3, 29, 29",
"c5.12xlarge":  "3, 29, 29",
"c5.18xlarge":  "3, 49, 49",
"c5.24xlarge":  "3, 49, 49",
"c5.metal":     "3, 49, 49",
"c5a.large":    "3, 9, 9",
"c5a.xlarge":   "3, 14, 14",
"c5a.2xlarge":  "3, 14, 14",
"c5a.4xlarge":  "3, 29, 29",
"c5a.8xlarge":  "3, 29, 29",
"c5a.12xlarge": "3, 29, 29",
"c5a.16xlarge": "3, 49, 49",
"c5a.24xlarge": "3, 49, 49",
"c5d.large":    "3, 9, 9",
"c5d.xlarge":   "3, 14, 14",
"c5d.2xlarge":  "3, 14, 14",
"c5d.4xlarge":  "3, 29, 29",
"c5d.9xlarge":  "3, 29, 29",
"c5d.18xlarge": "3, 49, 49",
"c5d.metal":    "3, 49, 49"
}
```

This leads me to be relatively certain https://github.com/cilium/cilium/commit/cb292b05400e5d204af852150942493b52329197 is the cause beacuse the only place the primary IP isn't excluded already is here: https://github.com/cilium/cilium/blob/38dd27af2f6c8246b8dc92ebf43ccbe8fab432ea/pkg/aws/eni/node.go#L197-L202

**Comment by gandro** [comment]
I agree that cb292b0 is likely the culprit here. I'm not sure why I convinced myself at the time that I don't need to remove it from the linked computation. I'll try to reproduce this issue and will try to fix it.

**Comment by gandro** [comment]
This will be fixed in 1.9.7 and 1.8.10.

Attached is a snippet for the workaround for 1.9.6 and 1.8.9. It contains default limits adjusted to compensate for the off-by-one error. This can be added to the `cilium-config` ConfigMap: [eni-limits.txt](https://github.com/cilium/cilium/files/6436114/eni-limits.txt)
